include/convex_bodies/ellipsoid.h

```diff
@@ -0,0 +1,300 @@
+        // }
+        // L = lltOfA.matrixL();
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+        c = Point(dim);
+        c.set_to_origin();
+    }
+
+
+    Ellipsoid(MT& Ain, Point& center) : A(Ain), c(center) {
+        // Eigen::LLT<Eigen::MatrixXd> lltOfA(A); // compute the Cholesky decomposition of Ain
+        // if(lltOfA.info() == Eigen::NumericalIssue) {
+        //     throw std::runtime_error("Possibly non semi-positive definitie matrix!");
+        // }
+        // L = lltOfA.matrixL();
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+    }
+
+
+    // Constructor for copula ellipsoid
+    Ellipsoid(std::vector<std::vector<NT> >& Ain) {
+        dim = Ain.size();
+        A.resize(dim, dim);
+        for (unsigned int i = 0; i < Ain.size(); i++) {
+            for (unsigned int j = 0; j < Ain.size(); j++) {
+                A(i,j) = Ain[i][j];
+            }
+        }
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+        c = Point(dim);
+        c.set_to_origin();
+    }
+
+
+    VT eigenvals() const {
+        return eigen_values;
+    }
+
+
+    VT eigenvals_inv() const {
+        return eigen_values_inv;
+    }
+
+
+    VT eigenvals_inv_sqrt() const {
+        return eigen_values_inv_sqrt;
+    }
+
+
+    MT eigenvecs() const {
+        return eigen_vecs;
+    }
+
+
+    unsigned int dimensions() const {
+        return dim;
+    }
+
+
+    void print() {
+        std::cout << "Ellipse is in the form: (x-c)' A (x-c) <= 1 \n";
+        std::cout << "c = \n" << c.print();
+        std::cout << "A = \n" << A;
+    }
+
+
+
+    NT mat_mult(Point const& p) {
+        NT res = NT(0);
+        for (size_t i=0; i<dim; i++) {
+            for (size_t j = i; j<dim; j++) {
```

Comment:
This part of the code is quite old. Please, improve it, using Eigen matrix-vector multiplication.

Reply:
Modified the code using this as reference: https://stackoverflow.com/a/63749321